enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,407 @@
+* Extend the cpu_manager_state to include a new `guaranteedCpuSet` tracking the available CPUs.
+* When a guaranteed QoS container using whole CPUs is created, allocate CPUs from the `guarateedCpuSet` instead 
+of the `defaultCpuSet` and set the CPU affinity to match.
+
+The next piece is to let the Kubernetes Scheduler take the split shared and guaranteed CPUSets into account when
+scheduling pods to a particular node. Kubernetes currently has a single 
+[CPU resource](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu) 
+that can be specified for each container. A change is necessary to ensure the scheduler will not over-allocate 
+shared or guaranteed CPUs on a node, resulting in a pod that would be scheduled but fail to run.
+
+The solution is to introduce two new [Extended Resources](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources):
+* `openshift.io/shared-cpus`: equal to number CPUs in shared CPUSet
+* `openshift.io/guaranteed-cpus`: equal to number of CPUs in isolated CPUSet
+
+Kubelet will be updated to publish the number of millicores available for each of these resources, based on the
+initial size of the `defaultCpuSet` and `guaranteedCpuSet` respectively.
+
+The Kubernetes Scheduler will then manage the allocation of these new extended resources for each node and ensure that
+the shared and guaranteed CPUs are not over-allocated. It would be unreasonable to expect the user to specify
+these new extended resources for every container and to ensure that the resource matches the type of container
+(shared vs. guaranteed). To automate this, an admission hook will be added to the 
+[Kubernetes API Server in OpenShift](https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission), 
+which will mutate each pod definition as it is created:
+* For containers with guaranteed whole CPUs, add a new resource request/limit `openshift.io/guaranteed-cpus` 
+equal to the number of CPUs requested in millicore units.
+* For containers with non-guaranteed CPU requests (or guaranteed fractional CPU requests) add a new resource 
+request/limit `openshift.io/shared-cpus` equal to the number of CPUs requested in millicore units. Note that 
+one of the restrictions for extended resources is that the request/limit must match - in the case of a container 
+with differing CPU requests/limits, the CPU request value will be used - this will preserve existing scheduler
+behaviour as only the CPU requests are used to choose the node a pod will run on.
+
+Here are the CPU requests/limits from an example container that uses shared CPUs:
+```yaml
+requests:
+  cpu: 200m
+limits:
+  cpu: 400m
+```
+This would be mutated as follows:
+```yaml
+requests:
+  cpu: 200m
+  openshift.io/shared-cpus: 200
+limits:
+  cpu: 400m
+  openshift.io/shared-cpus: 200
```

Comment:
This is really confusing.
If the scheduler looks only at the `requests` isn't that enough to only mutate the `requests` field?

Reply:
I think it worth mentioning in the enhancement then.

Reply:
Will do